tangelo/linq/simulator_base.py

```diff
@@ -0,0 +1,400 @@
+# Copyright 2021 Good Chemistry Company.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""SimulatorBase class, base class to define various target simulators and abstracting their
+differences from the user. Users can define their own target simulator as a child class of SimulatorBase.
+Able to run noiseless and noisy simulations, leveraging the capabilities of different backends,
+quantum or classical.
+
+If the user provides a noise model, then a noisy simulation is run with n_shots
+shots. If the user only provides n_shots, a noiseless simulation is run, drawing
+the desired amount of shots. If the target backend has access to the statevector
+representing the quantum state, we leverage any kind of emulation available to
+reduce runtime (directly generating shot values from final statevector etc). If
+the quantum circuit contains a MEASURE instruction, it is assumed to simulate a
+mixed-state and the simulation will be carried by simulating individual shots
+(e.g a number of shots is required).
+
+Some backends may only support a subset of the above. This information is
+contained in the static method backend_info defined by each subclass target.
+"""
+
+import abc
+import math
+from collections import Counter
+
+import numpy as np
+from scipy import stats
+from bitarray import bitarray
+from openfermion.ops import QubitOperator
+
+from tangelo.linq import Gate, Circuit
+from tangelo.linq.helpers.circuits.measurement_basis import measurement_basis_gates
+
+
+class SimulatorBase(abc.ABC):
+
+    def __init__(self, n_shots=None, noise_model=None):
+        """Instantiate Simulator object.
+
+        Args:
+            n_shots (int): Number of shots if using a shot-based simulator.
+            noise_model: A noise model object assumed to be in the format
+                expected from the target backend.
+        """
+        self._source = "abstract"
```

Comment:
Do a quick check: I think this attribute is unused. It had to do with translation, and was used in case we pushed non-tangelo circuit to this class. I think it is now irrelevant.